graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/BuiltinFunctions.java

```diff
@@ -286,6 +297,204 @@ public Object absObject(VirtualFrame frame, Object object,
         }
     }
 
+    /**
+     * Common class for all() and any() operations, as their logic and behaviors are very similar.
+     */
+    abstract static class AllOrAnyNode extends PNodeWithContext {
+        enum NodeType {
+            ALL,
+            ANY
+        }
+
+        @Child private PyObjectIsTrueNode isTrueNode = PyObjectIsTrueNode.create();
+
+        final private LoopConditionProfile loopConditionProfile = LoopConditionProfile.create();
+
+        abstract boolean execute(Frame frame, Object storageObj, NodeType nodeType);
+
+        @Specialization
+        boolean doBoolSequence(VirtualFrame frame,
+                        BoolSequenceStorage sequenceStorage,
+                        NodeType nodeType) {
+            boolean[] internalArray = sequenceStorage.getInternalBoolArray();
+            int i = 0;
+
+            while (loopConditionProfile.profile(i < sequenceStorage.length())) {
+                if (nodeType == NodeType.ALL && !isTrueNode.execute(frame, internalArray[i])) {
+                    return false;
+                } else if (nodeType == NodeType.ANY && isTrueNode.execute(frame, internalArray[i])) {
+                    return true;
+                }
+                i++;
+            }
+
+            return nodeType == NodeType.ALL;
+        }
+
+        @Specialization
+        boolean doIntSequence(VirtualFrame frame,
+                        IntSequenceStorage sequenceStorage,
+                        NodeType nodeType) {
+            int[] internalArray = sequenceStorage.getInternalIntArray();
+            int i = 0;
+
+            while (loopConditionProfile.profile(i < sequenceStorage.length())) {
+                if (nodeType == NodeType.ALL && !isTrueNode.execute(frame, internalArray[i])) {
+                    return false;
+                } else if (nodeType == NodeType.ANY && isTrueNode.execute(frame, internalArray[i])) {
+                    return true;
+                }
+            }
+
+            return nodeType == NodeType.ALL;
+        }
+
+        @Specialization
+        boolean doGenericSequence(VirtualFrame frame,
+                        SequenceStorage sequenceStorage,
+                        NodeType nodeType,
+                        @Cached SequenceStorageNodes.LenNode lenNode) {
+            Object[] internalArray = sequenceStorage.getInternalArray();
+            int i = 0;
+            int seqLength = lenNode.execute(sequenceStorage);
+
+            while (loopConditionProfile.profile(i < seqLength)) {
+                if (nodeType == NodeType.ALL && !isTrueNode.execute(frame, internalArray[i])) {
+                    return false;
+                } else if (nodeType == NodeType.ANY && isTrueNode.execute(frame, internalArray[i])) {
+                    return true;
+                }
+            }
```

Comment:
Missing `i++`? Btw, you can keep those as `for` loops, you can call the profile in the condition part (`for (int i = 0; i < loopConditionProfile.profile(i < seqLength); i++`)